common/path.cpp

```diff
@@ -178,8 +231,107 @@ Path Path::join(const String &str, char separator) const {
 
 Path Path::join(const char *str, char separator) const {
 	Path temp(*this);
-	temp.joinInPlace(str, DIR_SEPARATOR);
+	temp.joinInPlace(str, separator);
 	return temp;
 }
 
+Common::Array<Common::String> Path::splitComponents() const {
+	Common::Array<Common::String> res;
+	Common::String cur;
+	for (uint i = 0; i < _str.size(); i++) {
+		if (_str[i] == ESCAPER) {
+			i++;
+			if (_str[i] == ESCAPE_SLASH)
+				cur += '/';
+			else if (_str[i] == ESCAPE_SEPARATOR) {
+				res.push_back(cur);
+				cur = "";
+			}
+		} else
+			cur += _str[i];
+	}
+
+	res.push_back(cur);
+
+	return res;
+}
+
+Path Path::punycodeDecode() const {
+	Common::Array<Common::String> c = splitComponents();
+	String res;
+
+	for (uint i = 0; i < c.size(); i++) {
+		res += escapePath(punycode_decodefilename(c[i]));
+		if (i + 1 < c.size())
+			res += DIR_SEPARATOR;
+	}
+
+	Path ret;
+	ret._str = res;
+	return ret;
+}
+
+static String getIdentifierComponent(const String& in) {
+	String part = punycode_decodefilename(in);
+	String res = "";
+	for (uint j = 0; j < part.size(); j++)
+		if (part[j] == '/')
+			res += ':';
+		else
+			res += part[j];
+	return res;
+}
+
+String Path::getIdentifierString() const {
+	Common::Array<Common::String> c = splitComponents();
+	String res;
+
+	for (uint i = 0; i < c.size(); i++) {
+		res += getIdentifierComponent(c[i]);
+		if (i + 1 < c.size())
+			res += DIR_SEPARATOR;
+	}
+
+	return res;
+}
+
+Path Path::punycodeEncode() const {
+	Common::Array<Common::String> c = splitComponents();
+	String res;
+
+	for (uint i = 0; i < c.size(); i++) {
+		res += escapePath(punycode_encodefilename(c[i]));
+		if (i + 1 < c.size())
+			res += DIR_SEPARATOR;
+	}
+
+	Path ret;
+	ret._str = res;
+	return ret;
+}
+
+bool Path::matchPattern(const Common::Path& pattern) const {
+	Common::Array<Common::String> c = splitComponents();
+	Common::Array<Common::String> cpat = pattern.splitComponents();
+
+	// Prevent wildcards from matching the directory separator.
+	if (c.size() != cpat.size())
+		return false;
+
+	for (uint i = 0; i < c.size(); i++) {
+		if (!getIdentifierComponent(c[i]).matchString(getIdentifierComponent(cpat[i]), true))
+			return false;
+	}
+
+	return true;
+}
+
+bool Path::IgnoreCaseAndMac_EqualsTo::operator()(const Path& x, const Path& y) const {
```

Comment:
Please add a brief comment on what "Mac" mean here, e.g. explain the matching principles.

Reply:
Done